sympy/sets/sets.py

```diff
@@ -534,6 +535,70 @@ def measure(self):
         """
         return self._measure
 
+    @property
+    def kind(self):
+        """
+        This is a method for Sets class and its subclasses such as FiniteSet,
+        EmptySet, etc.
+
+        Every instance of Set or its subclass will have kind of class
+        ``SetKind``.
+        Parameter of this SetKind will depend on arguments of that set
+
+        Parameter of SetKind can be ``NumberKind``, ``UndefinedKind``,
+        ``TupleKind`` etc.
+        if all arguments of expression are not same then parameter will be
+        ``UndefinedKind``
+
+        ``FiniteSet(Matrix([1, 2])).kind`` kind of this expression will be
+        ``SetKind`` with ``MatrixKind(NumberKind)``
+        as its parameter because argument of FiniteSet is ``Matrix([1, 2])``.
+        Kind of this expression will be ``SetKind(MatrixKind(NumberKind))``
+
+        ``FiniteSet(0, Matrix([1, 2])).kind`` kind of this expression will be
+        ``SetKind`` with ``UndefinedKind``
+        as it parameter because kind of both argument of FiniteSet i.e. ``0``,
+        ``Matrix([1, 2])`` are different.
+        Kind of this expression will be ``SetKind(UndefinedKind)``
+
+        ``ProductSet(FiniteSet(1, Matrix([1, 2])), Interval(1, 2)).kind`` kind
+        of this expression will be ``SetKind`` with
+        ``TupleKind(UndefinedKind, NumberKind)`` as its parameter because Since
+        set is ProductSet so elements of this set will
+        be tuple. ``UndefinedKind`` is parameter of TupleKind because 1st
+        argument of ProductSet is ``FiniteSet(1, Matrix([1, 2]))``
+        which give ``UndefinedKind`` as its parameter. 2nd argument of
+        ProductSet is ``Interval(1, 2)`` whose paramter will be
+        ``NumberKind``. So Kind of this expression ``SetKind(TupleKind
+        (UndefinedKind, NumberKind))``
+
+        kind of ``UniversalSet`` and ``EmptySet`` are ``SetKind(UndefinedKind)
+        `` and ``SetKind()`` respectively.
+
+        Examples
+        ========
+
+        >>> from sympy import Interval, Matrix, FiniteSet, EmptySet, ProductSet, PowerSet
+        >>> FiniteSet(Matrix([1, 2])).kind
+        SetKind(MatrixKind(NumberKind))
+        >>> FiniteSet(0, Matrix([1, 2])).kind
+        SetKind(UndefinedKind)
+        >>> Interval(1,2).kind
+        SetKind(NumberKind)
+        >>> EmptySet.kind
+        SetKind()
+        >>> ProductSet(FiniteSet(1, Matrix([1, 2])), Interval(1, 2)).kind
+        SetKind(TupleKind(UndefinedKind, NumberKind))
```

Comment:
It's good to have one example (there is already a better one above) showing UndefinedKind but otherwise we don't really want to list more examples of badly defined sets. Also it would be better to show what some of these sets are e.g.:
```python
In [41]: p = ProductSet(FiniteSet(1, 2), FiniteSet(3, 4))

In [42]: p
Out[42]: {1, 2} × {3, 4}

In [43]: list(p)
Out[43]: [(1, 3), (2, 3), (1, 4), (2, 4)]

In [44]: p.kind
Out[44]: SetKind(TupleKind(NumberKind, NumberKind))
```